doc/high_availability.md

```diff
@@ -1,110 +1,74 @@
+
+The number of controllers can be reduced by invoking `remove-machine` with a 
+controller machine ID, and increased by re-running the `enable-ha` command.
+
+### Dqlite
+
+Each controller is a Dqlite node. The `dbaccessor` worker on each controller is 
+responsible for maintaining the Dqlite cluster. When entering HA mode, the 
+`dbaccessor` worker will configure the local Dqlite node as a member of the 
+cluster.
+
+When starting Dqlite, the worker must bind it to an IP address on the local
+machine. It is a requirement that there is a unique local-cloud-scoped address
+for Dqlite to use. If there is no unique address, new nodes will not be joined 
+to the cluster until one can be determined. See _Controller Charm_ below.
+
+Beyond joining nodes to the cluster, Juju does not manage Dqlite node roles.
+This is handled within Dqlite itself. A cluster will have one leader, voters
+which are eligible to participate in leader elections, stand-bys and spares.
+
+Juju does not predicate any logic on node roles.
+
+If the number of controller instances is reduced to one, the `dbaccessor` 
+worker will recognise this scenario and proactively reconfigure the cluster to
+be constituted by the local node only.
+
+### Controller Charm
+
+The controller charm propagates binding information to the `dbaccessor` worker.
+It coordinates with other controller units via the `db-cluster` peer relation.
+If there are multiple potential bind addresses for a Dqlite node, the user must
+supply an endpoint binding for the relation using a space that ensures a unique
+IP address.
+
+### API Addresses for Agents
+
+When machines in the control plane change, agent configuration files are 
+re-written with usable API addresses from all controllers. Agents will try
+these addresses in random order when connecting, so that an inaccessible 
+controller just results in connection to another.
+
+The list of addresses supplied to agent configuration can be influenced by the
+`juju-ctrl-space` controller configuration article. This is supplied with a
+space name in order that agent-controller communication can be isolated to 
+specific networks.
+
+### API Addresses for Clients
+
+Each time the Juju client establishes a connection to the Juju controller, it
+is sent the current list of API addresses and updates these in the local store. 
+The client's first connection attempt is always to the last address that it 
+used successfully. Others are tried subsequently if required.
+
+Addresses used by clients are not influenced by the `juju-ctrl-space` 
+configuration.
+
+### Singular Workers
+
+Many workers, such as the `dbaccessor` worker above, run on all controller 
+nodes, but there are some workers that must have exactly one instance running. 
+An obvious example of this is a model's compute provisioner - we don't want 
+multiple actors attempting to start a cloud instance for a new machine.
+
+The controller that such singular workers run on is determined by the lease
```

Comment:
**suggestion:**  If this documentation exists, we should have a link to it. If not, we should add a comment

```md
<!-- TODO: add a link to lease subsystem documentation whenever it lands -->
```

Reply:
Some notes and a possible solution:

(1) such as the `dbaccessor` worker above

This sort of reference usually signals an information architecture problem. 

(2) "run on all controller nodes"

Below we also speak of "instance". And I'm guessing we really mean "unit". Whichever we pick, I'd be consistent and maybe go for "unit" -- in Juju, horizontal scaling and, piggybacking on that, HA, is a matter of units. (Even if all labels apply, the inconsistency is causing me to have to think harder than strictly necessary.)

(3) "singular workers" 

Seems like unnecessary jargon.

(4) "must have exactly one instance running" 

Confusing -- what could it even mean in the context of HA? I'd rephrase as "run on exactly one controller unit, where the unit is determined by the lease sub-system."

Overall suggestion:
Have a section "Workers involved in controller HA" or something like that, then mention both the `dbaccessor` worker and others, then comment (this can even be a subsection -- even if the content is just one line, it's logically a subtopic) on how they differ with respect to whether they run on all controller units or on exactly one controller unit, as determined by the lease sub-system.